add FxPool support

0xAplki opened the pull request:
This PR is for adding support for our custom pool called `FxPool` in the SOR. It contains implementing our own logic to integrate with Balancer's router.


## How to test
Unit tests:
Run `yarn test:only test/xaveFxPool.spec.ts`

Integration tests:
```
 Pre-requisite:
- Add a polygon ALCHEMY_URL in .env
- Change --fork-block-number to 38546978
- Change hardhat.config.ts chainId to 137
- Run polygon node on terminal: yarn run node
```

Run `yarn test:only test/xaveFxPool.integration.spec.ts` to run the tests

johngrantuk commented on test/lib/subgraphPoolDataService.ts:
```diff
@@ -3,6 +3,7 @@ import { PoolDataService, SubgraphPoolBase } from '../../src';
 import { getOnChainBalances } from './onchainData';
 import { Provider } from '@ethersproject/providers';
 
+// @todo?
```
You can add your pool specific fields here which I believe will be:
delta
epsilon
and fxRate

johngrantuk commented on test/xaveFxPool.spec.ts:
```diff
@@ -0,0 +1,265 @@
+                SwapTypes.SwapExactOut
+            );
+
+            console.log(amount);
```
Could you update this to test against your expected values?

johngrantuk commented on src/pools/xaveFxPool/fxPool.ts:
```diff
@@ -0,0 +1,249 @@
+        if (swapType === SwapTypes.SwapExactIn) {
+            return bnum(
+                formatFixed(
+                    poolPairData.balanceIn.mul(this.MAX_IN_RATIO).div(ONE),
```
Just a check - this is currently using the same limits as a weighted pool, e.g. can only trade a 1/3 of token balance in the pool. Is this correct?

0xAplki replied:
we decided to remove the limits. our fxpool logic has an alpha region where it halts when we can no longer swap based on our swap invariant. hope this works!

johngrantuk replied:
The limit should reflect the max amount that can be swapped for that pool, so e.g. for the WeightedPools the limit is a 1/3 of the token balance because that reflects the smart contract limit. 
At the moment you limit is set to the total balance of the pool. But based off your explanation it sounds like your pools will also have an upper limit based on the alpha region and this is what should be reflected here?

0xAplki replied:
we added a quick calculation for the alpha region now @johngrantuk. this includes the upper limit on each swap. hope this works!

johngrantuk commented:
Thanks for the PR. It looks like its mostly there. Could you please add an integration test similar to this: https://github.com/balancer-labs/balancer-sor/blob/dd4cc15c984a82578c1d858f263a17912158c712/test/gyroE.integration.spec.ts where we can test the actual results against a fork of one of the networks where you already have pools. Thanks!

0xAplki commented:
ready for review again @johngrantuk! i have some issues on the integration tests where it doesnt return any routes but it reads the fixtures that set there. i just commented it out just like the others. let me knownif this works. thanks! :)

0xAplki commented:
hi @johngrantuk  noticed you did a lot of changes so I synced them also to this branch. the integration test is working as well. hope we can merge soon :) thanks!

johngrantuk commented on test/lib/subgraphPoolDataService.ts:
```diff
@@ -43,6 +43,9 @@ const queryWithLinear = `
           lambda
+          delta
+          epsilon
+          fxRate
```
The Subgraph doesn't seem to have `fxRate` field and throws an error with this query?

0xAplki replied:
i was not aware of the change of the name while we are developing this. changed the name to `latestFXPrice`

it's appearing when i query:
```
{
 tokens (where : {symbol: "XSGD"}){
     symbol
     latestFXPrice
  }
}
```

hope this works :)

johngrantuk replied:
Unfortunately we can't query like that as we need to fetch all the pools/tokens for the SOR. Also it seems like you have added `latestFxPrice` to the main pools query in subgraphPoolDataService but that throws an error. Maybe you need to add something to the Subgraph?

johngrantuk commented on yarn.lock:
```diff
@@ -1591,6 +1591,16 @@ ajv@^6.10.0, ajv@^6.12.4:
     json-schema-traverse "^0.4.1"
     uri-js "^4.2.2"
 
+ajv@^6.5.5:
```
Not sure why the yarn.lock has any changes? Could you please revert these if possible. Thanks.

0xAplki replied:
reverted already!

0xAplki commented:
Answered your comments and added some new updates to this PR. hope we can merge soon! :)

johngrantuk commented:
> Answered your comments and added some new updates to this PR. hope we can merge soon! :) @johngrantuk

Thanks @0xAplki. Its just the comment on limits that remains I think?

0xAplki commented:
> > Answered your comments and added some new updates to this PR. hope we can merge soon! :) @johngrantuk
> 
> Thanks @0xAplki. Its just the comment on limits that remains I think?

Updated this PR with that calculation on limits. Thanks @johngrantuk! :)

johngrantuk commented on test/lib/subgraphPoolDataService.ts:
```diff
@@ -43,6 +43,9 @@ const queryWithLinear = `
           lambda
+          delta
+          epsilon
+          latestFXPrice
```
This still doesn't seem to exit on SG?

0xAplki replied:
fixed!

johngrantuk commented on src/pools/xaveFxPool/fxPool.ts:
```diff
@@ -0,0 +1,276 @@
+
+    // There will be no limit amount to swap.
+    getLimitAmountSwap(
+        poolPairData: FxPoolPairData,
```
There's a bunch of warning being shown for mismatch in poolPairData, e.g.:
```
Property 'getLimitAmountSwap' in type 'FxPool' is not assignable to the same property in base type 'PoolBase<PoolPairBase>'.
  Type '(poolPairData: FxPoolPairData, swapType: SwapTypes) => BigNumber' is not assignable to type '(poolPairData: PoolPairBase, swapType: SwapTypes) => BigNumber'.
    Types of parameters 'poolPairData' and 'poolPairData' are incompatible.
      Type 'PoolPairBase' is not assignable to type 'FxPoolPairData'.
```
Could you please update to fix those?

0xAplki replied:
does this still exist on your end? i made some changes and it does appear already

johngrantuk commented on test/xaveFxPool.integration.spec.ts:
```diff
@@ -0,0 +1,142 @@
+dotenv.config();
+
+let sor: SOR;
+const networkId = Network.POLYGON;
```
I noticed you have some mainnet pools. Would it be possible to update this to use mainnet, (this already exists in our CI)?

0xAplki replied:
we dont have much there yet. most of our liquidity is in polygon atm. what would be the effect if we dont?

0xAplki replied:
added a mainnet implementation already :)
